src/Resolver/ValueResolver/ValueResolver.php

```diff
@@ -104,4 +110,29 @@ public function resolve(Expr $expr, Scope $scope, $unknownValuePlaceholder = nul
             return null;
         }
     }
+
+    public function resolveAsConstantType(Expr $expr, Scope $scope): ?Type
+    {
+        $values = $this->resolve($expr, $scope);
+        if ($values === null) {
+            return null;
+        }
+        $types = [];
+        foreach ($values as $value) {
+            if (is_string($value)) {
+                $types[] = new ConstantStringType($value);
+            } elseif (is_int($value)) {
+                $types[] = new ConstantIntegerType($value);
+            } elseif (is_float($value)) {
+                $types[] = new ConstantFloatType($value);
+            } elseif (is_bool($value)) {
+                $types[] = new ConstantBooleanType($value);
+            } elseif (is_null($value)) {
+                $types[] = new NullType();
+            } else {
```

Comment:
So if one of 5 values can’t be covered we silently return null? Imho we should return union type of 4/5 precisio or throw exception

Reply:
Or get type by scope :)

Reply:
We use null as "cannot be resolved" in value resolvers. I would rather avoid exception for non-exceptional result of resolving.

Reply:
Yes thats why I wrote to get type from scope